File: blf-ts/src/blam/halo4/v20810_12_09_22_1647_main/game/megalogamengine/megalogamengine_variable_metadata.ts
import type {
  c_bitstream_reader,
  c_bitstream_writer,
} from "../../../../../bitstream";
import { AutoMap } from "../../../../../helpers/automap";
import { e_multiplayer_team_designator } from "../game_engine_default";
import { c_custom_variable_reference } from "./megalogamengine_references";

/** Variable replication mode (`network-state`, 2 bits). Halo 4 count=3. */
export enum e_megalo_variable_network_state {
  local = 0,
  networked = 1,
  networked_high = 2,
}

/** Halo 4 `VariableFlags`: network-state + is-persistent. */
export class s_megalo_variable_flags {
  @AutoMap(() => Number)
  m_network_state: e_megalo_variable_network_state =
    e_megalo_variable_network_state.local;
  @AutoMap(() => Boolean)
  m_is_persistent = false;
  decode(bitstream: c_bitstream_reader): void {
    this.m_network_state = bitstream.read_enum(
      "network-state",
      2,
      e_megalo_variable_network_state
    );
    this.m_is_persistent = bitstream.read_bool("is-persistent");
  }
  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_enum(
      this.m_network_state,
      2,
      e_megalo_variable_network_state
    );
    bitstream.write_bool(this.m_is_persistent);
  }
}

export class s_variable_metadata {
  @AutoMap(() => [c_custom_variable_reference])
  m_numeric_variables: [
    c_custom_variable_reference,
    s_megalo_variable_flags,
  ][] = [];
  @AutoMap(() => [c_custom_variable_reference])
  m_timer_variables: c_custom_variable_reference[] = [];
  @AutoMap(() => [Number])
  m_team_variables: [e_multiplayer_team_designator, s_megalo_variable_flags][] =
    [];
  @AutoMap(() => [s_megalo_variable_flags])
  m_player_variables: s_megalo_variable_flags[] = [];
  @AutoMap(() => [s_megalo_variable_flags])
  m_object_variables: s_megalo_variable_flags[] = [];
  constructor(
    private readonly numeric_variable_count_bits: number,
    private readonly timer_variable_count_bits: number,
    private readonly team_variable_count_bits: number,
    private readonly player_variable_count_bits: number,
    private readonly object_variable_count_bits: number
  ) {}
  decode(bitstream: c_bitstream_reader): void {
    const numeric_variable_count = bitstream.read_integer(
      "numeric-variable-count",
      this.numeric_variable_count_bits
    );
    for (let i = 0; i < numeric_variable_count; i++) {
      const numeric_variable = new c_custom_variable_reference();
      numeric_variable.decode(bitstream);
      const flags = new s_megalo_variable_flags();
      flags.decode(bitstream);
      this.m_numeric_variables.push([numeric_variable, flags]);
    }
    const timer_variable_count = bitstream.read_integer(
      "timer-count",
      this.timer_variable_count_bits
    );
    for (let i = 0; i < timer_variable_count; i++) {
      const timer_variable = new c_custom_variable_reference();
      timer_variable.decode(bitstream);
      this.m_timer_variables.push(timer_variable);
    }
    const team_variable_count = bitstream.read_integer(
      "team-variable-count",
      this.team_variable_count_bits
    );
    for (let i = 0; i < team_variable_count; i++) {
      const team_variable_value = bitstream.read_enum(
        "team-variable-value",
        4,
        e_multiplayer_team_designator
      );
      const flags = new s_megalo_variable_flags();
      flags.decode(bitstream);
      this.m_team_variables.push([team_variable_value, flags]);
    }
    const player_variable_count = bitstream.read_integer(
      "player-variable-count",
      this.player_variable_count_bits
    );
    for (let i = 0; i < player_variable_count; i++) {
      const flags = new s_megalo_variable_flags();
      flags.decode(bitstream);
      this.m_player_variables.push(flags);
    }
    const object_variable_count = bitstream.read_integer(
      "object-variable-count",
      this.object_variable_count_bits
    );
    for (let i = 0; i < object_variable_count; i++) {
      const flags = new s_megalo_variable_flags();
      flags.decode(bitstream);
      this.m_object_variables.push(flags);
    }
  }
  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_integer(
      this.m_numeric_variables.length,
      this.numeric_variable_count_bits
    );
    for (const [numeric_variable, flags] of this.m_numeric_variables) {
      numeric_variable.encode(bitstream);
      flags.encode(bitstream);
    }
    bitstream.write_integer(
      this.m_timer_variables.length,
      this.timer_variable_count_bits
    );
    for (const timer_variable of this.m_timer_variables) {
      timer_variable.encode(bitstream);
    }
    bitstream.write_integer(
      this.m_team_variables.length,
      this.team_variable_count_bits
    );
    for (const [team_variable, flags] of this.m_team_variables) {
      bitstream.write_enum(team_variable, 4, e_multiplayer_team_designator);
      flags.encode(bitstream);
    }
    bitstream.write_integer(
      this.m_player_variables.length,
      this.player_variable_count_bits
    );
    for (const flags of this.m_player_variables) {
      flags.encode(bitstream);
    }
    bitstream.write_integer(
      this.m_object_variables.length,
      this.object_variable_count_bits
    );
    for (const flags of this.m_object_variables) {
      flags.encode(bitstream);
    }
  }
}
export function s_variable_metadata_global(): s_variable_metadata {
  // MegaloLimits: num20, timer8, team8, player10, object18
  return new s_variable_metadata(5, 4, 4, 4, 5);
}
export function s_variable_metadata_player(): s_variable_metadata {
  // MegaloLimits: num10, timer4, team4, player4, object4
  return new s_variable_metadata(4, 3, 3, 3, 3);
}
export function s_variable_metadata_object(): s_variable_metadata {
  // MegaloLimits: num10, timer4, team2, player4, object6
  return new s_variable_metadata(4, 3, 2, 3, 3);
}
export function s_variable_metadata_team(): s_variable_metadata {
  // MegaloLimits: num10, timer4, team4, player4, object6
  return new s_variable_metadata(4, 3, 3, 3, 3);
}

File: blf_lib/src/blam/halo4/v20810_12_09_22_1647_main/game/megalogamengine/megalogamengine_variable_metadata.rs
use serde::{Deserialize, Serialize};
use num_derive::{FromPrimitive, ToPrimitive};
use blf_lib::blam::halo4::v20810_12_09_22_1647_main::game::game_engine_team::e_multiplayer_team_designator;
use blf_lib::blam::halo4::v20810_12_09_22_1647_main::game::megalogamengine::megalogamengine_custom_variable_reference::c_custom_variable_reference;
use blf_lib::io::bitstream::{c_bitstream_reader, c_bitstream_writer};
use blf_lib_derivable::result::BLFLibResult;

/// Variable replication mode (`network-state`, 2 bits). Halo 4 count=3.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default, ToPrimitive, FromPrimitive, crate::derive::c_enum)]
#[bits(2)]
pub enum e_megalo_variable_network_state {
    #[default]
    local = 0,
    networked = 1,
    networked_high = 2,
}

/// Halo 4 `VariableFlags`: network-state + is-persistent.
#[derive(Default, PartialEq, Debug, Clone, Copy, Serialize, Deserialize)]
pub struct s_megalo_variable_flags {
    pub m_network_state: e_megalo_variable_network_state,
    pub m_is_persistent: bool,
}

impl s_megalo_variable_flags {
    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        bitstream.write_enum(self.m_network_state)?;
        bitstream.write_bool(self.m_is_persistent)?;
        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_network_state = bitstream.read_enum("network-state")?;
        self.m_is_persistent = bitstream.read_bool("is-persistent")?;
        Ok(())
    }
}

#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct s_variable_metadata <
    const numeric_variable_count_bits: usize,
    const timer_variable_count_bits: usize,
    const team_variable_count_bits: usize,
    const player_variable_count_bits: usize,
    const object_variable_count_bits: usize,
> {
    pub m_numeric_variables: Vec<(c_custom_variable_reference, s_megalo_variable_flags)>,
    pub m_timer_variables: Vec<c_custom_variable_reference>,
    pub m_team_variables: Vec<(e_multiplayer_team_designator, s_megalo_variable_flags)>,
    pub m_player_variables: Vec<s_megalo_variable_flags>,
    pub m_object_variables: Vec<s_megalo_variable_flags>,

}

impl<
    const numeric_variable_count_bits: usize,
    const timer_variable_count_bits: usize,
    const team_variable_count_bits: usize,
    const player_variable_count_bits: usize,
    const object_variable_count_bits: usize,
>
s_variable_metadata<
    numeric_variable_count_bits,
    timer_variable_count_bits,
    team_variable_count_bits,
    player_variable_count_bits,
    object_variable_count_bits
>
{
    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        bitstream.write_integer(self.m_numeric_variables.len() as u8, numeric_variable_count_bits)?;
        for (numeric_variable, flags) in self.m_numeric_variables.iter() {
            numeric_variable.encode(bitstream)?;
            flags.encode(bitstream)?;
        }

        bitstream.write_integer(self.m_timer_variables.len() as u8, timer_variable_count_bits)?;
        for timer_variable in self.m_timer_variables.iter() {
            timer_variable.encode(bitstream)?;
        }

        bitstream.write_integer(self.m_team_variables.len() as u8, team_variable_count_bits)?;
        for (team_variable, flags) in self.m_team_variables.iter() {
            bitstream.write_enum(*team_variable)?;
            flags.encode(bitstream)?;
        }

        bitstream.write_integer(self.m_player_variables.len() as u8, player_variable_count_bits)?;
        for flags in self.m_player_variables.iter() {
            flags.encode(bitstream)?;
        }

        bitstream.write_integer(self.m_object_variables.len() as u8, object_variable_count_bits)?;
        for flags in self.m_object_variables.iter() {
            flags.encode(bitstream)?;
        }

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        let numeric_variable_count: u16 = bitstream.read_integer("numeric-variable-count", numeric_variable_count_bits)?;
        for _ in 0..numeric_variable_count {
            let mut numeric_variable = c_custom_variable_reference::default();
            numeric_variable.decode(bitstream)?;
            let mut flags = s_megalo_variable_flags::default();
            flags.decode(bitstream)?;
            self.m_numeric_variables.push((numeric_variable, flags));
        }

        let timer_variable_count: u16 = bitstream.read_integer("timer-variable-count", timer_variable_count_bits)?;
        for _ in 0..timer_variable_count {
            let mut timer_variable = c_custom_variable_reference::default();
            timer_variable.decode(bitstream)?;
            self.m_timer_variables.push(timer_variable);
        }

        let team_variable_count: u16 = bitstream.read_integer("team-variable-count", team_variable_count_bits)?;
        for _ in 0..team_variable_count {
            let team_variable_value = bitstream.read_enum("team-variable-value")?;
            let mut flags = s_megalo_variable_flags::default();
            flags.decode(bitstream)?;
            self.m_team_variables.push((team_variable_value, flags));
        }

        let player_variable_count: u16 = bitstream.read_integer("player-variable-count", player_variable_count_bits)?;
        for _ in 0..player_variable_count {
            let mut flags = s_megalo_variable_flags::default();
            flags.decode(bitstream)?;
            self.m_player_variables.push(flags);
        }

        let object_variable_count: u16 = bitstream.read_integer("object-variable-count", object_variable_count_bits)?;
        for _ in 0..object_variable_count {
            let mut flags = s_megalo_variable_flags::default();
            flags.decode(bitstream)?;
            self.m_object_variables.push(flags);
        }

        Ok(())
    }
}
